File: Benchmarks/TriangleCount/C++/trianglecount.cpp
#include <algorithm>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <iterator>
#include <sstream>
#include <string>
#include <unordered_set>
#include <vector>

using namespace std;

vector<string> splitString(const string &str, char delimiter) {
  stringstream stream(str);
  vector<string> split;
  string current;

  while (getline(stream, current, delimiter)) {
    split.push_back(current);
  }

  return split;
}

class Graph {
public:
  int vertices;
  vector<unordered_set<int>> edges;

public:
  Graph(const string &path) {
    ifstream file(path);
    string line;

    while (getline(file, line)) {
      if (line[0] == 'p') {
        vector<string> values = splitString(line, ' ');
        int nvertices = atoi(values[2].c_str());

        this->vertices = nvertices;
        this->edges = vector<unordered_set<int>>(nvertices);
      } else if (line[0] == 'a' || line[0] == 'e') {
        vector<string> values = splitString(line, ' ');

        int v0 = atoi(values[1].c_str());
        int v1 = atoi(values[2].c_str());

        if (v0 != v1) {
          edges[v0].insert(v1);
          edges[v1].insert(v0);
        }
      }
    }

    file.close();
  }
};

int TriangleCount_single(const Graph &graph, int a, int b) {
  if (a >= b)
    return 0;

  auto edgesA = graph.edges[a];
  auto edgesB = graph.edges[b];

  vector<int> intersection;
  set_intersection(edgesA.begin(), edgesA.end(), edgesB.begin(), edgesB.end(),
                   back_inserter(intersection));

  return intersection.size();
}

int TriangleCount(const Graph &graph) {
  int count = 0;

  for (int a = 0; a < graph.vertices; a++) {
    for (int b = a + 1; b < graph.vertices; b++) {
      count += TriangleCount_single(graph, a, b);
    }
  }

  return count;
}

int main(int argc, const char **argv) {
  string file = argv[1];

  Graph graph(file);
  auto count = TriangleCount(graph);

  std::cout << count << std::endl;

  return 0;
}

File: Benchmarks/TriangleCount/Java/TriangleCount.java

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashSet;

class Graph {

    public int nvertices;
    public ArrayList<HashSet<Integer>> edges;

    public Graph(String file) throws FileNotFoundException, IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {

            String line = reader.readLine();

            while (line != null) {
                if (line.startsWith("p")) {
                    String[] values = line.split(" ");
                    int vertices = Integer.parseInt(values[2]);

                    this.nvertices = vertices;
                    this.edges = new ArrayList<>(vertices);
                    for (int i = 0; i < vertices; i++) {
                        this.edges.add(new HashSet<>());
                    }
                } else if (line.startsWith("e") || line.startsWith("a")) {
                    String[] values = line.split(" ");

                    int a = Integer.parseInt(values[1]);
                    int b = Integer.parseInt(values[2]);

                    if (a != b) {
                        this.edges.get(a).add(b);
                        this.edges.get(b).add(a);
                    }
                }

                line = reader.readLine();
            }
        }
    }
}

class TriangleCount {

    public static int TriangleCount_single(Graph graph, int a, int b) {
        if (a >= b)
            return 0;

        HashSet<Integer> edgesA = graph.edges.get(a);
        HashSet<Integer> edgesB = graph.edges.get(b);

        HashSet<Integer> intersection = new HashSet<>(edgesA);
        intersection.retainAll(edgesB);

        return intersection.size();

    }

    public static int Count(Graph graph) {
        int count = 0;

        for (int a = 0; a < graph.nvertices; a++) {
            for (int b = a + 1; b < graph.nvertices; b++) {
                count += TriangleCount_single(graph, a, b);
            }
        }

        return count;
    }

    public static void main(String[] args) throws FileNotFoundException, IOException {
        String file = args[0];

        Graph graph = new Graph(file);
        int count = Count(graph);

        System.out.println(count);
    }
}
